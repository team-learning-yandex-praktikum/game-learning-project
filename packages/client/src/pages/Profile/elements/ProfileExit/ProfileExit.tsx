import { FC, useCallback } from 'react'
import styles from './profileExit.module.css'
import Title from '@components/Title'
import Button from '@components/Button'
import { useNavigate } from 'react-router'
import { Routes } from '@routes/constants'
import { useAppDispatch } from '@store/hooks'
import { logout } from '@store/user'

export const ProfileExit: FC = () => {
    const navigate = useNavigate()
    const dispatch = useAppDispatch()

    const handleClick = useCallback(async () => {
        try {
            await dispatch(logout()).unwrap()
            navigate(Routes.Login)
        } catch (e) {
            console.error(e)
        }
    }, [dispatch, navigate])

    return (
        <>
            <Title className={styles.title}>Уходите?</Title>
            <Button
                className={styles.button}
                variant="outlined"
                onClick={handleClick}
                color="secondary"
            >
                Да
            </Button>
        </>
    )
}

export default ProfileExit
