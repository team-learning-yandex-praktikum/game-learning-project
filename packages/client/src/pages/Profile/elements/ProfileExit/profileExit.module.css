.title {
  margin-bottom: 100px;
  color: var(--color-accent-secondary);
  composes: font-design-headline-l from global;
}
