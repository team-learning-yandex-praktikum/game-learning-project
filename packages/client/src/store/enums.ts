export enum LoadStatus {
    idle = 'idle',
    loading = 'loading',
    complete = 'complete',
    failed = 'failed',
}
