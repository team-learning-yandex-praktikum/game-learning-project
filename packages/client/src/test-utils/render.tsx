import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@store'
import { ReactNode } from 'react'

export const withProviders = (component: ReactNode) => {
    const testStore = store

    return (
        <BrowserRouter>
            <Provider store={testStore}>{component}</Provider>
        </BrowserRouter>
    )
}
