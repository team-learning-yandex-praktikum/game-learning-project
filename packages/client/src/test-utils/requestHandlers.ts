import { http, HttpResponse } from 'msw'
import { joinURL } from '@utils/url/joinURL'
import { BASE_API } from '@api/baseApi/constants'
import { UserDTO } from '@api/auth/types'
import { transformData } from '@utils'
import { UpdatingProfileDTO } from '@api/user/types'

export const userInfoStub: UserDTO = {
    id: 111,
    login: 'ivan',
    display_name: null,
    avatar: null,
    first_name: 'Иван',
    second_name: 'Иванов',
    email: 'ivan@test.ru',
    phone: '79999999999',
}

export const requestHandlers = {
    signup: http.post(joinURL(BASE_API, 'auth/signup'), () =>
        HttpResponse.text(String(userInfoStub.id))
    ),
    signin: http.post(joinURL(BASE_API, 'auth/signin'), () =>
        HttpResponse.text('OK')
    ),
    me: http.get(joinURL(BASE_API, 'auth/user'), () =>
        HttpResponse.json(userInfoStub)
    ),
    logout: http.post(joinURL(BASE_API, 'auth/logout'), () =>
        HttpResponse.text('OK')
    ),
    updatePassword: http.put(joinURL(BASE_API, 'user/password'), () =>
        HttpResponse.text('OK')
    ),
    updateProfile: http.put(
        joinURL(BASE_API, 'user/profile'),
        async ({ request }) => {
            const body = (await request.json()) as UpdatingProfileDTO
            return HttpResponse.json(
                transformData.from.dto({
                    ...userInfoStub,
                    ...body,
                })
            )
        }
    ),
}
