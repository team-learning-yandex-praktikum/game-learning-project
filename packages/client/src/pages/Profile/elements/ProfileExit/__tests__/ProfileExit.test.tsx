import { setupServer } from 'msw/node'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import ProfileExit from '../ProfileExit'
import { withProviders } from '../../../../../test-utils/render'
import userEvent from '@testing-library/user-event'
import { requestHandlers } from '../../../../../test-utils/requestHandlers'
import { store } from '@store'
import { LoadStatus } from '@store/enums'

const server = setupServer(requestHandlers.logout)

describe('Profile Exit', () => {
    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())

    test('correctly rendered', () => {
        const { asFragment } = render(withProviders(<ProfileExit />))
        expect(asFragment()).toMatchSnapshot()
    })

    test('should change state after exit', async () => {
        render(withProviders(<ProfileExit />))
        const exitButton = screen.getByText('Да')

        await userEvent.click(exitButton)

        const userState = store.getState().user

        expect(userState?.data).toStrictEqual({})
        expect(userState?.status).toBe(LoadStatus.complete)
    })
})
