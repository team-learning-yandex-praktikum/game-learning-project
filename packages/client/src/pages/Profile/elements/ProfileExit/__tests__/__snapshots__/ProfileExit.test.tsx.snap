// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Exit correctly rendered 1`] = `
<DocumentFragment>
  <h2
    class=""
  >
    Уходите?
  </h2>
  <button
    class=""
  >
    Да
  </button>
</DocumentFragment>
`;
